pkg/koordlet/runtimehooks/reconciler/reconciler.go

```diff
@@ -41,36 +42,147 @@ const (
 	ContainerLevel ReconcilerLevel = "container"
 )
 
-// map[string]*cgroupReconciler: key is cgroup filename,
-var globalCgroupReconciler = map[ReconcilerLevel]map[string]*cgroupReconciler{
-	KubeQOSLevel:   {},
-	PodLevel:       {},
-	ContainerLevel: {},
+var globalCgroupReconcilers = struct {
+	all []*cgroupReconciler
+
+	kubeQOSLevel   map[string]*cgroupReconciler
+	podLevel       map[string]*cgroupReconciler
+	containerLevel map[string]*cgroupReconciler
+}{
+	kubeQOSLevel:   map[string]*cgroupReconciler{},
+	podLevel:       map[string]*cgroupReconciler{},
+	containerLevel: map[string]*cgroupReconciler{},
 }
 
 type cgroupReconciler struct {
 	cgroupFile  system.CgroupFile
-	fn          reconcileFunc
 	description string
+	level       ReconcilerLevel
+	filter      Filter
+	fn          map[string]reconcileFunc
+}
+
+// Filter & Conditions:
+// 1. a condition for one cgroup file should have no more than one filter/index func
+// 2. different indexes of one cgroup file can have different reconcile functions
+// 3. indexes for one cgroup should be enumerable
+type Filter interface {
+	Name() string
+	Filter(podMeta *statesinformer.PodMeta) string
+}
+
+type noneFilter struct{}
+
+const (
+	NoneFilterCondition = ""
+	NoneFilterName      = "none"
+)
+
+func (d *noneFilter) Name() string {
+	return NoneFilterName
+}
+
+func (d *noneFilter) Filter(podMeta *statesinformer.PodMeta) string {
+	return NoneFilterCondition
+}
+
+var singletonNoneFilter *noneFilter
+
+// NoneFilter returns a Filter which skip filtering anything (into the same condition)
+func NoneFilter() *noneFilter {
+	if singletonNoneFilter == nil {
+		singletonNoneFilter = &noneFilter{}
+	}
+	return singletonNoneFilter
+}
+
+type podQOSFilter struct{}
+
+const (
+	PodQOSFilterName = "podQOS"
+)
+
+func (p *podQOSFilter) Name() string {
+	return PodQOSFilterName
+}
+
+func (p *podQOSFilter) Filter(podMeta *statesinformer.PodMeta) string {
+	return string(apiext.GetPodQoSClass(podMeta.Pod))
+}
+
+var singletonPodQOSFilter *podQOSFilter
+
+// PodQOSFilter returns a Filter which filters pod qos class
+func PodQOSFilter() *podQOSFilter {
+	if singletonPodQOSFilter == nil {
+		singletonPodQOSFilter = &podQOSFilter{}
+	}
+	return singletonPodQOSFilter
 }
 
 type reconcileFunc func(protocol.HooksProtocol) error
 
-func RegisterCgroupReconciler(level ReconcilerLevel, cgroupFile system.CgroupFile,
-	fn reconcileFunc, description string) {
-	if _, ok := globalCgroupReconciler[level]; !ok {
-		klog.Fatalf("resource level %v has not init", level)
+func RegisterCgroupReconciler(level ReconcilerLevel, cgroupFile system.CgroupFile, description string,
```

Comment:
we need some comments about how to use this method

Reply:
done